### ISSUE TITLE
Add TemporaryFolder.getRoot()

### ISSUE THREAD

**Issue by ChristianCiach** [glytching/junit-extensions#8]
I am in the process of converting a lot of unit tests to JUnit 5. Almost all of these that use the old TemporaryFolder rule use getRoot() to get the temporary folder itself.

With your extensions, I've changed all these calls to `TemporaryFolder.createDirectory("")` to mimic this behaviour, but this seems hacky.

Also, I think it's a bit surprising that your Extension can only create a single instance of TemporaryFolder. If unit tests define multiple TemporaryFolder-parameters, all of them will be the same instance.

**Comment by glytching** [comment]
Thanks for raising the issue.

I'm open to making some changes here but I think I need some additional details so that I can understand the rationale for them.

>  I've changed all these calls to TemporaryFolder.createDirectory("") to mimic this behaviour, but this seems hacky

If you want to get access to the root folder so that you can create a file or directory within the root directory then you should be using `createFile()`, `createDirectory()`. However, you've already seen that `createDirectory()` meets that need so I'm guessing that you have some other use in mind for the root directory. Could you explain - maybe by providing an example - why you want to get access to the root folder?

> I think it's a bit surprising that your Extension can only create a single instance of TemporaryFolder. If unit tests define multiple TemporaryFolder-parameters, all of them will be the same instance.

Are you proposing something like this?

```  
  @Test
  @ExtendWith(TemporaryFolderExtension.class)
  public void canInjectATemporaryFolderAsAParameter(TemporaryFolder thisTemporaryFolder, TemporaryFolder thatTemporaryFolder)
      throws IOException {
    ...
  }
```

If so, I hadn't really considered the possibility that someone would want to inject multiple instances of `TemporaryFolder` rather than, say, injecting one and calling `createDirectory()` multiple times. Could you explain - maybe by providing an example - why you might want to inject multiple instances of `TemporaryFolder`.

**Comment by ChristianCiach** [comment]
The "why" is the same for both questions: I need do convert several hundred of unit tests that all use getRoot() :) So to make it easy, I've just replaced all getRoot() calls with createDirectory(""). Also, I don't see why I should be forced to create a sub-directory inside a directory that is already temporary.

Most of the unit tests test classes that accept a java.io.File as a (constructor-)parameter, als almost all of these tests just pass tmpFolder.getRoot().

**Comment by ChristianCiach** [comment]
And yes, your example of using multiple TemporaryFolder parameters is exactly what I needed. For me, this is not important anymore, as only a handful of unit tests needed more than one temporary folder, and I could manually edit these to just call temporaryFolder.createDirectory("bla[1-2]").

**Comment by glytching** [comment]
Ok, so your tests do something like 

```
doSomething(temporaryFolder.getRoot());
// ... assert
```
And the equivalent - using this extension - is:

```
doSomething(temporaryFolder.createDirectory(someRandomDirectoryName));
// ... assert
```

So, this extension - in a bid to hide the root folder, since the  extension is responsible for creating and deleting it - is (a) too verbose for your use case and (b) creates a sub directory when you only need to use the root directory.

If that assessment is correct then I can see that exposing `getRoot()` would help but I still have concerns about the second issue. The `TemporaryFolderExtension` could be re-worked to inject (and manage) _N_ instances of `TemporaryFolder` but I wonder whether this is a sufficiently general use case to warrant the additional complexity in`TemporaryFolder`. Perhaps a public accessor such as `anyDirectory()` would reduce the verbosity (there would be no need to create `someRandomDirectoryName`) a little bit whilst still using a single instance of `TemporaryFolder`. The caller would be forced to invoke a method other than `getRoot()` but there wouldn't be any material issue with creating additional folders since the net increase in folder creation is only _one_; create one instance of `TemporaryFolder` with multiple sub folders vs. create multiple instances of `TemporaryFolder`.

So your test might look like:

```
// when a single instance of temporaryFolder (and hence the same root folder) suffices
doSomething(temporaryFolder.getRoot());

// when a single instance of temporaryFolder (and hence the same root folder) does not suffice
doSomething(temporaryFolder.anyDirectory());
doSomething(temporaryFolder.anyDirectory());
```

Thoughts?

**Comment by ChristianCiach** [comment]
Well, yeah, in theory I agree with everything you said. 

In my case, even a single unit test may call getRoot() several times, and of course expect the same result with every call. So instead of `temporaryFolder.createDirectory(someRandomDirectoryName)` I've opted for `temporaryFolder.createDirectory("")`, just because it's shorter :) This also does return the root folder itself, but this of course isn't strictly necessary.

Pragmatically speaking, your library advertises itself with offering replacements for popular JUnit 4 Rules, so I think it is safe to assume that many people will choose this library in an attempt to convert legacy tests to JUnit 5. Because of this, I think it would be the right thing to offer a very similar API as the legacy JUnit 4 Rules to ease the migration. As it was possible to define multiple TemporaryFolder-Rules in a single unit test with JUnit 4, I see no reason why your library should differ here. (And if it does, it should at least throw an exception or something to prevent misuse). 

Your proposal of `temporaryFolder.anyDirectory()` would certainly help for tests that originally used multiple TemporaryFolder-Rules, but again, this makes refactoring of several hundreds to thousands unit tests (as in my case) a lot harder than just converting multiple legacy Rules to multiple TemporaryFolder-parameters.

**Comment by ChristianCiach** [comment]
Keep in mind that there is no need for you (of course there is not "need" anyway) to do anything here, as I was able to archive everything I needed to do. These are just my thoughts on things that could have been nicer, because these where the things I stumpled upon while migrating A LOT of unit tests to JUnit 5. Let me use this opportunity to **thank you** for this library which made my life just a bit easier :)

**Comment by glytching** [comment]
I'm adding `TemporaryFolder.getRoot()`, in addition to your specific use case the argument for 'ease of migration' is compelling.

Thanks for raising the issue.

**Comment by ChristianCiach** [comment]
Thanks :) I absolutely agree that it is safer to not expose the root folder, but well, JUnit 4 did... And of course, any malicious caller could just call `createDirectory("../bla")` anyway (or just call `createDirectory("")` as I did, which kinda feels malicious).

**Comment by glytching** [comment]
Yeah, so much for my safety conciousness ... I didn't like exposing the `rootFolder` but didn't notice the `createDirectory("")` backdoor :)

Anyway, mirroring the analogous JUnit4 behaviour for 'ease of migration' (as long as that behaviour isn't definitely undesireable) makes sense.

Thanks again, much appreciated.

**Comment by glytching** [comment]
I have added `TemporaryFolder.getRoot()` and published a new version of this library - 2.2.0 - to [Maven Central](http://repo1.maven.org/maven2/io/github/glytching/junit-extensions/2.2.0/).
